Generate genotype df

KatalinaBobowik opened the pull request:
This generates a genotype file to Plink format to be used in the association analysis. Only  variants with an MAF > 0.01 are retained + additional filtering for passing VQSR filters, GQ, and missingness. This has run [successfully](https://batch.hail.populationgenomics.org.au/batches/8390/jobs/2) in the `test` bucket.

Commit:
First commit

Commit:
Export output as PLINK format and warp script in function to launch through the analysis runner

Commit:
Add in README

Commit:
Describe reason as to why I filter at 0.8 call rate

MattWellie commented on scripts/generate_genotype_df/generate_genotype_df.py:
```diff
@@ -0,0 +1,31 @@
+
+    mt = hl.read_matrix_table(TOB_WGS)
+    # filter out variants that didn't pass the VQSR filter
+    mt = mt.filter_rows(hl.is_missing(mt.filters))
```
I thought Hail used an empty set for 0 applied filters? I've just tried this syntax in a notebook and all rows with no applied filters were removed.

This alternate syntax allows filters to be missing or a 0 length set, and will pass in both situations

```
matrix = matrix.filter_rows(
    hl.len(
        hl.or_else(
            matrix.filters, 
            hl.empty_set(hl.tstr)
        )
    ) == 0
)
```

vladsavelyev replied:
In the joint-calling pipeline we have empty sets for column-level filters, and NA for row-level filters. Good catch - should unify that. 

But agree that it's better to always handle both situations using the code you shared. Thanks Matt!

KatalinaBobowik replied:
that's really helpful, thanks @MattWellie!

vladsavelyev reviewed (approved):
Just +1 to Matt's suggestion, approving otherwise

lgruen commented on scripts/generate_genotype_df/generate_genotype_df.py:
```diff
@@ -0,0 +1,31 @@
+    mt = mt.filter_rows(mt.freq.AF[1] > 0.01)
+    # export as PLINK file
+    tob_wgs_path = output_path('tob_wgs_plink_maf01')
+    hl.export_plink(mt, tob_wgs_path, ind_id=mt.s)
```
As discussed over Zoom, would it be an option to export the Hail Table as a Pandas DataFrame directly, e.g. using Parquet?

KatalinaBobowik replied:
I've exported this to parquet now

lgruen commented on scripts/generate_genotype_df/main.py:
```diff
@@ -0,0 +1,23 @@
+    batch,
+    f'generate_genotype_df.py',
+    max_age='4h',
+    num_secondary_workers=20,
```
Do you need that many workers and the disk override or is this just a remnant from a previous script?

KatalinaBobowik replied:
This was an old remnant- thanks for pointing it out!

Commit:
Add in suggestion from Matt Welland and decrease n workers

As per Matt Wellands suggestion, I updated filtering the `mt.filters` to handle cases where filters can be missing or a 0 length set. I also decreased the number of secondary workers to the deafult (the 20 workers was a holdover from teh last script) and I exported data to parquet, which will make reading in the genotype file much quicker.

lgruen commented on scripts/generate_genotype_df/README.md:
```diff
@@ -0,0 +1,9 @@
+# Export TOB-WGS dataset to PLINK format
+
+This runs a Hail query script in Dataproc using Hail Batch in order to export the TOB-WGS genotypes, which are requirements for the association analysis. Only variants that pass VQSR and GQ filters are retained, as well as variants with an MAF > 0.01. To run, use conda to install the analysis-runner, then execute the following command:
```
Nit: the analysis-runner is now installed using pip.

KatalinaBobowik replied:
Fixed

lgruen commented on scripts/generate_genotype_df/generate_genotype_df.py:
```diff
@@ -0,0 +1,32 @@
+    mt = mt.filter_rows(hl.len(hl.or_else(mt.filters, hl.empty_set(hl.tstr))) == 0)
+
+    # VQSR does not filter out low quality genotypes. Filter these out
+    mt = mt.filter_entries(mt.GQ <= 20, keep = False)
```
```suggestion
    mt = mt.filter_entries(mt.GQ <= 20, keep=False)
```

lgruen commented on scripts/generate_genotype_df/generate_genotype_df.py:
```diff
@@ -0,0 +1,32 @@
+    mt = mt.filter_rows(hl.agg.sum(hl.is_missing(mt.GT)) > (n_samples * call_rate), keep=False)
+    # filter out variants with MAF < 0.01
+    mt = mt.filter_rows(mt.freq.AF[1] > 0.01)
+    # export as PLINK file
```
```suggestion
```

lgruen commented on scripts/generate_genotype_df/generate_genotype_df.py:
```diff
@@ -0,0 +1,32 @@
+    mt = mt.filter_rows(mt.freq.AF[1] > 0.01)
+    # export as PLINK file
+    tob_wgs_path = output_path('tob_wgs_maf01.parquet')
+    mt.rows().to_spark().write.mode("append").parquet(tob_wgs_path)
```
This will write a Spark DataFrame, not a Pandas DataFrame. Are you sure that's what you need downstream?

KatalinaBobowik replied:
This saves just fine as a parquet file that I can load in as a pandas df with `pd.read_parquet`. Is there a better way to do this?

lgruen replied:
There's `to_pandas`, maybe that will have a slightly different format? @vladsaveliev, do you know?

KatalinaBobowik replied:
ok, I tested this out and while both dfs are the same when reading back into python, the `to_spark` way results in a much larger file. I've now changed this to `to_pandas` (and only take the fields I need), and then export to parquet

lgruen replied:
Cool, thanks! You probably also don't need the "append" write mode here, right?

KatalinaBobowik replied:
No, I just did the `mt.rows().to_pandas().to_parquet()`

lgruen replied:
Perfect! 😸

lgruen commented on scripts/generate_genotype_df/generate_genotype_df.py:
```diff
@@ -0,0 +1,32 @@
+    mt = mt.filter_rows(mt.freq.AF[1] > 0.01)
+    # export as PLINK file
+    tob_wgs_path = output_path('tob_wgs_maf01.parquet')
+    mt.rows().to_spark().write.mode("append").parquet(tob_wgs_path)
```
```suggestion
    mt.rows().to_spark().write.mode("append").parquet(tob_wgs_path)
```
```suggestion
    mt.rows().to_spark().write.mode('append').parquet(tob_wgs_path)
```

lgruen commented on scripts/generate_genotype_df/main.py:
```diff
@@ -0,0 +1,21 @@
+    f'generate_genotype_df.py',
+    max_age='4h',
+    init=['gs://cpg-reference/hail_dataproc/install_common.sh'],
+    job_name=f'export_genotype_data',
```
```suggestion
    job_name='export_genotype_data',
```

lgruen commented on scripts/generate_genotype_df/main.py:
```diff
@@ -0,0 +1,21 @@
+
+dataproc.hail_dataproc_job(
+    batch,
+    f'generate_genotype_df.py',
```
```suggestion
    'generate_genotype_df.py',
```

Commit:
Update scripts/generate_genotype_df/generate_genotype_df.py

Co-authored-by: Leonhard Gruenschloss <leonhard.gruenschloss@populationgenomics.org.au>

Commit:
Update scripts/generate_genotype_df/generate_genotype_df.py

Co-authored-by: Leonhard Gruenschloss <leonhard.gruenschloss@populationgenomics.org.au>

Commit:
Update scripts/generate_genotype_df/generate_genotype_df.py

Co-authored-by: Leonhard Gruenschloss <leonhard.gruenschloss@populationgenomics.org.au>

Commit:
Update scripts/generate_genotype_df/main.py

Co-authored-by: Leonhard Gruenschloss <leonhard.gruenschloss@populationgenomics.org.au>

Commit:
Update scripts/generate_genotype_df/main.py

Co-authored-by: Leonhard Gruenschloss <leonhard.gruenschloss@populationgenomics.org.au>

Commit:
Update README

lgruen reviewed (approved):
Added a note regarding `to_spark` vs `to_pandas`, which you might want to consider changing before merging.

Commit:
Export file to pandas before saving as parquet file, since this results in a smaller file than to_spark
